modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -3348,6 +3348,25 @@ def broadcast_apply_full_axis(
                         kw["column_widths"] = self._column_widths_cache
                     elif len(new_columns) == 1 and new_partitions.shape[1] == 1:
                         kw["column_widths"] = [1]
+        else:
+            if (
+                kw["row_lengths"] is None
```

Comment:
```suggestion
                axis == 0 and
                kw["row_lengths"] is None
```

It seems that we can preserve lengths only for one axis (for the one, that we're applying a full-axis function). Consider this case, where kernel modifies the shapes of partitions without modifying the total length:
```python
from modin.test.storage_formats.pandas.test_internals import construct_modin_df_by_scheme
import pandas

md_df = construct_modin_df_by_scheme(pandas.DataFrame({"a": [1, 2, 3, 4], "b": [3, 4, 5, 6], "c": [6, 7, 8, 9], "d": [0, 1, 2, 3]}), {"row_lengths": [2, 2], "column_widths": [2, 2]})._query_compiler._modin_frame

def func(df):
    if df.iloc[0, 0] == 1:
        return pandas.DataFrame({"a": [1, 2, 3], "b": [2, 3, 4], "c": [4, 2, 1], "d": [3, 2, 1]})
    else:
        return pandas.DataFrame({"a": [3], "b": [4], "c": [1], "d": [1]})

res = md_df.apply_full_axis(func=func, axis=1, new_index=[0, 1, 2, 3], new_columns=["a", "b", "c", "d"], keep_partitioning=True)
print(res._row_lengths_cache) # [2, 2]
print(res._column_widths_cache) # [2, 2]

actual_row_lengths = [part.length() for part in res._partitions[:, 0]]
print(actual_row_lengths) # [3, 1]
```